src/tray/mod.rs

```diff
@@ -0,0 +1,424 @@
+            0,
+            0,
+            0,
+            0,
+            0,
+            ptr::null_mut(),
+            ptr::null_mut(),
+            h_instance,
+            &mut tray_state as *mut _ as *mut std::ffi::c_void,
+        );
+
+        // Send window handle back to main loop so we can post messages to it
+        let _ = TRAY_SENDER
+            .get()
+            .unwrap()
+            .try_send(TrayMessage::TrayCreated(WindowHandle(hwnd)));
+
+        let mut msg = mem::zeroed();
+        while GetMessageW(&mut msg, ptr::null_mut(), 0, 0) != 0 {
+            TranslateMessage(&msg);
+            DispatchMessageW(&msg);
+        }
+
+        if tray_state.close_gui {
+            // WM_QUIT was sent, which was triggered by the Exit button. Close entire program.
+            SHOULD_EXIT.store(true, Ordering::Relaxed);
+
+            // Activate window to force event loop to run / see that it should now close
+            ShowWindow(tray_state.gui_handle.unwrap(), SW_SHOWMINIMIZED);
+        }
+    });
+}
+
+unsafe fn add_icon(hwnd: HWND) {
+    let h_instance = GetModuleHandleW(ptr::null());
+
+    let icon_handle = LoadIconW(h_instance, MAKEINTRESOURCEW(0x101));
+
+    let mut tooltip_array = [0u16; 128];
+    let tooltip = "Ajour";
+    let mut tooltip = tooltip.encode_utf16().collect::<Vec<_>>();
+    tooltip.extend(vec![0; 128 - tooltip.len()]);
+    tooltip_array.swap_with_slice(&mut tooltip[..]);
+
+    let mut icon_data: NOTIFYICONDATAW = mem::zeroed();
+    icon_data.cbSize = mem::size_of::<NOTIFYICONDATAW>() as u32;
+    icon_data.hWnd = hwnd;
+    icon_data.uID = 1;
+    icon_data.uCallbackMessage = WM_APP;
+    icon_data.uFlags = NIF_ICON | NIF_MESSAGE | NIF_TIP;
+    icon_data.hIcon = icon_handle;
+    icon_data.szTip = tooltip_array;
+
+    Shell_NotifyIconW(NIM_ADD, &mut icon_data);
+}
+
+unsafe fn display_balloon_message(hwnd: HWND, message: &str) {
+    let mut info = [0u16; 256];
+    let message = str_to_wide!(message);
+
+    for (idx, b) in message[0..message.len().min(255)].iter().enumerate() {
+        info[idx] = *b;
+    }
+
+    let mut icon_data: NOTIFYICONDATAW = mem::zeroed();
+    icon_data.cbSize = mem::size_of::<NOTIFYICONDATAW>() as u32;
+    icon_data.hWnd = hwnd;
+    icon_data.uID = 1;
+    icon_data.uFlags = NIF_INFO;
+    icon_data.szInfo = info;
+    icon_data.dwInfoFlags = NIIF_NONE | NIIF_NOSOUND;
+
+    Shell_NotifyIconW(NIM_MODIFY, &mut icon_data);
+}
+
+unsafe fn remove_icon(hwnd: HWND) {
+    let mut icon_data: NOTIFYICONDATAW = mem::zeroed();
+    icon_data.cbSize = mem::size_of::<NOTIFYICONDATAW>() as u32;
+    icon_data.hWnd = hwnd;
+    icon_data.uID = 1;
+
+    Shell_NotifyIconW(NIM_DELETE, &mut icon_data);
+}
+
+unsafe fn show_popup_menu(hwnd: HWND, state: &TrayState) {
+    let menu = CreatePopupMenu();
+
+    let hidden = state.gui_hidden;
+
+    let mut about = str_to_wide!("About...");
+    let mut toggle = str_to_wide!(if hidden { "Show window" } else { "Hide window" });
```

Comment:
We I think we should order and name the `popup_menu` like:

```
Show Ajour
About
----------
Exit
```
Also wondering if theres an API to make a horizontal line as seen on a few trays?

![image](https://user-images.githubusercontent.com/2248455/113978941-e5785a00-9844-11eb-89dc-dfede4bada78.png)

Reply:
Good idea! Let me look into this